server/polar/order/service.py

```diff
@@ -1242,5 +1244,46 @@ async def _on_order_paid(self, session: AsyncSession, order: Order) -> None:
                 subscription_id=order.subscription_id,
             )
 
+    async def handle_payment_failure(
+        self, session: AsyncSession, order: Order
+    ) -> Order:
+        """Handle payment failure for an order, initiating dunning if necessary."""
+        if order.subscription is None:
+            return order
+
+        if order.subscription.stripe_subscription_id is not None:
```

Comment:
Yes, that's the solution for that. Quite simple, but effective IMO. I applied the same approach when looking for subscriptions to cycle:

https://github.com/polarsource/polar/blob/3bd15bced0a977829c68a9840493c48e132766cd/server/polar/subscription/scheduler.py#L40-L53